pkg/util/validation/limits_map.go

```diff
@@ -0,0 +1,92 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package validation
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"gopkg.in/yaml.v3"
+)
+
+// LimitsMap is a generic map that can hold either float64 or int64 as values.
+type LimitsMap[T float64 | int64] struct {
+	data      map[string]T
+	validator func(k string, v T) error
+}
+
+func NewLimitsMap[T float64 | int64](validator func(k string, v T) error) LimitsMap[T] {
+	return LimitsMap[T]{
+		data:      make(map[string]T),
+		validator: validator,
+	}
+}
+
+// String implements flag.Value
+func (m LimitsMap[T]) String() string {
+	out, err := json.Marshal(m.data)
+	if err != nil {
+		return fmt.Sprintf("failed to marshal: %v", err)
+	}
+	return string(out)
+}
+
+// Set implements flag.Value
+func (m LimitsMap[T]) Set(s string) error {
+	newMap := make(map[string]T)
+	if err := json.Unmarshal([]byte(s), &newMap); err != nil {
+		return err
+	}
+	return m.updateMap(newMap)
+}
+
+// Clone returns a copy of the LimitsMap.
+func (m LimitsMap[T]) Clone() LimitsMap[T] {
+	newMap := make(map[string]T, len(m.data))
+	for k, v := range m.data {
+		newMap[k] = v
+	}
+	return LimitsMap[T]{data: newMap, validator: m.validator}
+}
+
+// UnmarshalYAML implements yaml.Unmarshaler.
+func (m LimitsMap[T]) UnmarshalYAML(value *yaml.Node) error {
+	newMap := make(map[string]T)
+	if err := value.DecodeWithOptions(newMap, yaml.DecodeOptions{KnownFields: true}); err != nil {
+		return err
+	}
+	return m.updateMap(newMap)
+}
+
+func (m LimitsMap[T]) updateMap(newMap map[string]T) error {
+	for k, v := range newMap {
+		if m.validator != nil {
+			if err := m.validator(k, v); err != nil {
+				return err
+			}
+		}
+
+		m.data[k] = v
+	}
```

Comment:
I think we should validate all fields first, then update.

Reply:
Good shout. I ported over the implementation we had before, but I agree with you that not allowing partial updates seems like a better approach - this is actually a breaking change (or a bug fix?) from the way the rate limit limit works so I'll have to add to the readme. Can you confirm that this assumption is correct?

Done in https://github.com/grafana/mimir/pull/8378/commits/e5440806807ea215073a31fac4bd4275ee57fedf

Reply:
I think it's a minor bug that might not even be worth mentioning in the changelog. The fix lgtm.